File: problem-solving/_lc-674-easy-longest-continuous-increasing-subsequence/2-dp-linear-time.py
"""
https://leetcode.com/problems/longest-continuous-increasing-subsequence/description/

Complexity
----------
Time: O(n)
Space: O(n)
"""

class Solution:
    def findLengthOfLCIS(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [1] * n
        for i in range(1, n):
            if nums[i - 1] < nums[i]:
                dp[i] = dp[i - 1] + 1
        return max(dp)

File: problem-solving/_lc-674-easy-longest-continuous-increasing-subsequence/2-dp-linear-time.go
/*
https://leetcode.com/problems/longest-continuous-increasing-subsequence/

Complexity
----------
Time: O(n)
Space: O(n)
*/

func findLengthOfLCIS(nums []int) int {
    n := len(nums)
    dp := make([]int, n)

    // Initialise each with 1
    for i := 0; i < n; i++ {
        dp[i] = 1
    }
    // Solve using DP
    for i := 1; i < n; i++ {
        if nums[i - 1] < nums[i] {
            dp[i] = dp[i - 1] + 1
        }
    }
    // Find max
    res := dp[0]
    for i := 1; i < n; i++ {
        if res < dp[i] {
            res = dp[i]
        }
    }
    return res
}
